### DISCUSSION TITLE
How we can get the negotiated TLS version?

### DISCUSSION THREAD

**Discussion by afshinpir** [openssl/openssl#24916] [Q&A] [answered]
Hi,

If I create my SSL context using version flexible method(`TLS_server_method()`, `TLS_client_method()`), how I can get the negotiated TLS version after connection establishment?

In addition, is the negotiated version should be the highest version supported by both side or it is a random selection between supported versions? I means for example if I connect a tls client and server which both has only set:
```
SSL_CTX_set_min_proto_version(m_ctx, TLS1_2_VERSION);
```
should I always get TLS1.3 if both sides support it? Because I had experienced getting both TLS1.2 and TLS1.3 connection on same code which I found strange.

**Comment by t8m** [comment, marked as answer]
Use SSL_version() or SSL_get_version() functions to obtain the negotiated protocol version. However this must be called only after the handshake is completed to get the correct result. Did you ensure that? That might be the reason for seeing TLS-1.2 when TLS-1.3 was actually negotiated.

**Comment by LyraHarring** [comment]
You can check the negotiated [TLS version](https://cheapsslweb.com/blog/list-of-tls-versions-which-is-the-latest-tls-version/) after the handshake using SSL_get_version() or SSL_get_protocol_version() on the established SSL* object. The negotiated version is not random—it’s chosen as the highest version supported by both client and server. For example, if you set SSL_CTX_set_min_proto_version(..., TLS1_2_VERSION) on both ends, you will only see TLS 1.2, even if TLS 1.3 is available. If you want TLS 1.3, you must allow it by not restricting the max version. Hope it helps!
